### PR TITLE
Add API Gateway policies

### PR DESCRIPTION
**Pull request by chrsmith**
Adds three new policies for API Gateway, as part of #13.

### CODE REVIEW

File: integration-tests/apiGateway/index.ts [hunk 1]
```diff
@@ -0,0 +1,109 @@
+// Copyright 2016-2019, Pulumi Corporation.
```
Comment (ekrengel) [inline integration-tests/apiGateway/index.ts]: 2020! Although I guess you did open this in 2019 🤷‍♀

File: src/apiGateway.ts [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+            }
+            if (!methodSettings.settings.cacheDataEncrypted) {
+                reportViolation(`API Gateway Method '${methodSettings.methodPath}' not configured to encrypt cached responses.`);
+            }
+        }),
+    };
+}
+
+export interface ApiGatewayEndpointTypeCheckArgs extends PolicyArgs {
+    /** Whether or not API Endpoint type EDGE is allowed. */
```
Comment (ekrengel) [inline src/apiGateway.ts]: Should these comments include what it defaults to?

File: src/apiGateway.ts [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    allowEdge?: boolean;
+    /** Whether or not API Endpoint type REGIONAL is allowed. */
+    allowRegional?: boolean;
+    /** Whether or not API Endpoint type PRIVATE is allowed. */
+    allowPrivate?: boolean;
+}
+
+/** @internal */
+export function apiGatewayEndpointTypeCheck(
```
Comment (ekrengel) [inline src/apiGateway.ts]: Should we get rid of the "check" here and below? I believe thats what we did for others.

File: integration-tests/apiGateway/index.ts (hunk 1 above)
Comment (chrsmith) [reply integration-tests/apiGateway/index.ts]: 🤷‍♂  ... I'll change it to 2020.

File: src/apiGateway.ts [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+/** @internal */
+export function apiGatewayStageCached(enforcementLevel?: EnforcementLevel): ResourceValidationPolicy {
+    return {
+        name: "apigateway-stage-cached",
+        description: "Checks that API Gateway Stages have a cache cluster enabled.",
+        enforcementLevel: enforcementLevel || defaultEnforcementLevel,
+        validateResource: validateTypedResource(aws.apigateway.Stage, (stage, _, reportViolation) => {
+            if (!stage.cacheClusterEnabled) {
+                reportViolation(`API Gateway Stage '${stage.stageName}' does not have a cache cluster enabled.`);
```
[the remark this answers is not in the document]
Comment (chrsmith) [reply src/apiGateway.ts]: Thanks, done.

File: src/apiGateway.ts [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    return {
+        name: "apigateway-method-cached-and-encrypted",
+        description: "Checks API Gateway Methods that responses are configured to be cached and that those cached responses are encrypted.",
+        enforcementLevel: enforcementLevel || defaultEnforcementLevel,
+        validateResource: validateTypedResource(aws.apigateway.MethodSettings, (methodSettings, _, reportViolation) => {
+            if (!methodSettings.settings.cachingEnabled) {
+                reportViolation(`API Gateway Method '${methodSettings.methodPath}' does not have caching enabled.`);
+            }
+            if (!methodSettings.settings.cacheDataEncrypted) {
+                reportViolation(`API Gateway Method '${methodSettings.methodPath}' not configured to encrypt cached responses.`);
```
[the remark this answers is not in the document]
Comment (chrsmith) [reply src/apiGateway.ts]: Done, and also updated the string above it to have the form "must have..."

File: src/apiGateway.ts [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+        description: "Checks API Gateway endpoint configuration is one of the allowed types.",
+        enforcementLevel: enforcementLevel || defaultEnforcementLevel,
+        validateResource: validateTypedResource(aws.apigateway.RestApi, (restApi, _, reportViolation) => {
+            let endpointType = "(endpointConfiguration.types unspecified)";
+            if (restApi.endpointConfiguration) {
+                endpointType = restApi.endpointConfiguration.types;
+            }
+
+            const wrappedReportViolation = () => {
+                reportViolation(`API Gateway '${restApi.name}' has an unsupported endpoint type '${endpointType}'.`);
```
[the remark this answers is not in the document]
Comment (chrsmith) [reply src/apiGateway.ts]: Sounds good. It's perhaps unlikely we can make it concise, but we can at least make the long error message clear.

File: src/apiGateway.ts [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+                        wrappedReportViolation();
+                    }
+                    break;
+                case "PRIVATE":
+                    if (!allowPrivate) {
+                        wrappedReportViolation();
+                    }
+                    break;
+
+                default:
```
[the remark this answers is not in the document]
Comment (chrsmith) [reply src/apiGateway.ts]: Done.

File: src/apiGateway.ts [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+
+/** @internal */
+export function apiGatewayMethodCachedAndEncrypted(enforcementLevel?: EnforcementLevel): ResourceValidationPolicy {
+    return {
+        name: "apigateway-method-cached-and-encrypted",
+        description: "Checks API Gateway Methods that responses are configured to be cached and that those cached responses are encrypted.",
+        enforcementLevel: enforcementLevel || defaultEnforcementLevel,
+        validateResource: validateTypedResource(aws.apigateway.MethodSettings, (methodSettings, _, reportViolation) => {
+            if (!methodSettings.settings.cachingEnabled) {
+                reportViolation(`API Gateway Method '${methodSettings.methodPath}' does not have caching enabled.`);
```
[the remark this answers is not in the document]
Comment (chrsmith) [reply src/apiGateway.ts]: done.

File: src/apiGateway.ts (hunk 2 above)
Comment (chrsmith) [reply src/apiGateway.ts]: Yes, done.

File: src/apiGateway.ts (hunk 3 above)
Comment (chrsmith) [reply src/apiGateway.ts]: Yeah, looks like it. Done.

**Pull request merged**